### ISSUE TITLE
About the metrics for loss and accuracy

### ISSUE THREAD

**Issue by leeyunhome** [ultralytics/yolov5#2469]
## ❔Question

checked with wandb
The loss and accuracy graph is shown below.

I don’t think I’m learning something well
What can you be suspicious of in this case?

The image size is set to 320.
![image](https://user-images.githubusercontent.com/59910227/111096578-361cd000-8583-11eb-84f5-1ab66dcf92e3.png)
![image](https://user-images.githubusercontent.com/59910227/111096601-446aec00-8583-11eb-827f-344758ea407e.png)

![image](https://user-images.githubusercontent.com/59910227/111096749-914ec280-8583-11eb-9282-a22f9d1dde5b.png)

**Comment by glenn-jocher** [comment]
@leeyunhome hey there. I think your code may have a problem, as multiple runs are being logged to the same run. It's out of date also, as current master shows epochs in your W&B console, not steps.

Your first training seems fine, but too short. I would first update your code (clone a new version of the repo or git pull to update your current), and then simply train 2x longer than your first run.

**Comment by leeyunhome** [comment]
> @leeyunhome hey there. I think your code may have a problem, as multiple runs are being logged to the same run. It's out of date also, as current master shows epochs in your W&B console, not steps.
> 
> Your first training seems fine, but too short. I would first update your code (clone a new version of the repo or git pull to update your current), and then simply train 2x longer than your first run.

Hello,

Do you mean double the epoch and try learning from scratch?

Thank you.

**Comment by glenn-jocher** [comment]
@leeyunhome yes, your first training seems fine but too short. Double the epochs and retrain for better val loss and mAP. I'll post our full guide for improving results below.

👋 Hello! Thanks for asking about improving training results. Most of the time good results can be obtained with no changes to the models or training settings, **provided your dataset is sufficiently large and well labelled**. If at first you don't get good results, there are steps you might be able to take to improve, but we always recommend users first train with all default settings before considering any changes. This helps establish a performance baseline and spot areas for improvement.

If you have questions about your training results **we recommend you provide the maximum amount of information possible** if you expect a helpful response, including results plots (train losses, val losses, P, R, mAP), PR curve, confusion matrix, training mosaics, test results and dataset statistics images such as labels.png. All of these are located in your `project/name` directory, typically `yolov5/runs/train/exp`.

We've put together a full guide for users looking to get the best results on their YOLOv5 trainings below.

## Dataset

- **Images per class.** ≥1.5k images per class
- **Instances per class.** ≥10k instances (labeled objects) per class total
- **Image variety.** Must be representative of deployed environment. For real-world use cases we recommend images from different times of day, different seasons, different weather, different lighting, different angles, different sources (scraped online, collected locally, different cameras) etc.
- **Label consistency.** All instances of all classes in all images must be labelled. Partial labelling will not work.
- **Label accuracy.** Labels must closely enclose each object. No space should exist between an object and it's bounding box. No objects should be missing a label.
- **Background images.** Background images are images with no objects that are added to a dataset to reduce False Positives (FP). We recommend about 0-10% background images to help reduce FPs (COCO has 1000 background images for reference, 1% of the total).

<a href="https://arxiv.org/abs/1405.0312"><img width="800" src="https://user-images.githubusercontent.com/26833433/109398377-82b0ac00-78f1-11eb-9c76-cc7820669d0d.png" alt="COCO Analysis"></a>


## Model Selection

Larger models like YOLOv5x will produce better results in nearly all cases, but have more parameters and are slower to run. For mobile applications we recommend YOLOv5s/m, for cloud or desktop applications we recommend YOLOv5l/x. See our README [table](https://github.com/ultralytics/yolov5#pretrained-checkpoints) for a full comparison of all models. 

To start training from pretrained weights simply pass the name of the model to the `--weights` argument. Models download automatically from the [latest YOLOv5 release](https://github.com/ultralytics/yolov5/releases).
```bash
python train.py --data custom.yaml --weights yolov5s.pt
                                             yolov5m.pt
                                             yolov5l.pt
                                             yolov5x.pt
```

<p align="center"><img width="700" alt="YOLOv5 Models" src="https://user-images.githubusercontent.com/26833433/103595982-ab986000-4eb1-11eb-8c57-4726261b0a88.png"></p>

## Training Settings

Before modifying anything, **first train with default settings to establish a performance baseline**. A full list of train.py settings can be found in the [train.py](https://github.com/ultralytics/yolov5/blob/master/train.py) argparser.

- **Epochs.** Start with 300 epochs. If this overfits early then you can reduce epochs. If overfitting does not occur after 300 epochs, train longer, i.e. 600, 1200 etc epochs.
- **Image size.** COCO trains at native resolution of `--img 640`, though due to the high amount of small objects in the dataset it can benefit from training at higher resolutions such as `--img 1280`. If there are many small objects then custom datasets will benefit from training at native or higher resolution. Best inference results are obtained at the same `--img` as the training was run at, i.e. if you train at `--img 1280` you should also test and detect at `--img 1280`.
- **Batch size.** Use the largest `--batch-size` that your hardware allows for. Small batch sizes produce poor batchnorm statistics and should be avoided.
- **Hyperparameters.** Default hyperparameters are in [hyp.scratch.yaml](https://github.com/ultralytics/yolov5/blob/master/data/hyp.scratch.yaml). We recommend you train with default hyperparameters first before thinking of modifying any. In general, increasing augmentation hyperparameters will reduce and delay overfitting, allowing for longer trainings and higher final mAP. Reduction in loss component gain hyperparameters like `hyp['obj']` will help reduce overfitting in those specific loss components. For an automated method of optimizing these hyperparameters, see our [Hyperparameter Evolution Tutorial](https://docs.ultralytics.com/yolov5/tutorials/hyperparameter_evolution).

## Further Reading

If you'd like to know more a good place to start is Karpathy's 'Recipe for Training Neural Networks', which has great ideas for training that apply broadly across all ML domains:
http://karpathy.github.io/2019/04/25/recipe/

**Comment by leeyunhome** [comment]
> @leeyunhome hey there. I think your code may have a problem, as multiple runs are being logged to the same run. It's out of date also, as current master shows epochs in your W&B console, not steps.
> 
> Your first training seems fine, but too short. I would first update your code (clone a new version of the repo or git pull to update your current), and then simply train 2x longer than your first run.

Hello,

First, I am re-learning with existing code without git pull.
Image size 320, 400 epochs rotated,
Currently, 14 epochs are in progress.

![image](https://user-images.githubusercontent.com/59910227/111108939-27dbad80-859d-11eb-960a-be126187763c.png)

Can I assume that learning is going well?

Thank you.

**Comment by glenn-jocher** [comment]
@leeyunhome yeah everything is going well in your plots, though I would recommend training with the latest code for best results.

**Comment by leeyunhome** [comment]
> @leeyunhome yes, your first training seems fine but too short. Double the epochs and retrain for better val loss and mAP. I'll post our full guide for improving results below.
> 
> 👋 Hello! Thanks for asking about improving training results. Most of the time good results can be obtained with no changes to the models or training settings, **provided your dataset is sufficiently large and well labelled**. If at first you don't get good results, there are steps you might be able to take to improve, but we always recommend users first train with all default settings before considering any changes. This helps establish a performance baseline and spot areas for improvement.
> 
> If you have questions about your training results **we recommend you provide the maximum amount of information possible** if you expect a helpful response, including results plots (train losses, val losses, P, R, mAP), PR curve, confusion matrix, training mosaics, test results and dataset statistics images such as labels.png. All of these are located in your `project/name` directory, typically `yolov5/runs/train/exp`.
> 
> We've put together a full guide for users looking to get the best results on their YOLOv5 trainings below.
> 
> ## Dataset
> * **Images per class.** ≥1.5k images per class
> * **Instances per class.** ≥10k instances (labeled objects) per class total
> * **Image variety.** Must be representative of deployed environment. For real-world use cases we recommend images from different times of day, different seasons, different weather, different lighting, different angles, different sources (scraped online, collected locally, different cameras) etc.
> * **Label consistency.** All instances of all classes in all images must be labelled. Partial labelling will not work.
> * **Label accuracy.** Labels must closely enclose each object. No space should exist between an object and it's bounding box. No objects should be missing a label.
> * **Background images.** Background images are images with no objects that are added to a dataset to reduce False Positives (FP). We recommend about 0-10% background images to help reduce FPs (COCO has 1000 background images for reference, 1% of the total).
> 
> [<img alt="COCO Analysis" width="800" src="https://user-images.githubusercontent.com/26833433/109398377-82b0ac00-78f1-11eb-9c76-cc7820669d0d.png">](https://arxiv.org/abs/1405.0312)
> 
> ## Model Selection
> Larger models like YOLOv5x will produce better results in nearly all cases, but have more parameters and are slower to run. For mobile applications we recommend YOLOv5s/m, for cloud or desktop applications we recommend YOLOv5l/x. See our README [table](https://github.com/ultralytics/yolov5#pretrained-checkpoints) for a full comparison of all models.
> 
> To start training from pretrained weights simply pass the name of the model to the `--weights` argument. Models download automatically from the [latest YOLOv5 release](https://github.com/ultralytics/yolov5/releases).
> 
> ```shell
> python train.py --data custom.yaml --weights yolov5s.pt
>                                              yolov5m.pt
>                                              yolov5l.pt
>                                              yolov5x.pt
> ```
> 
> <img alt="YOLOv5 Models" width="700" src="https://user-images.githubusercontent.com/26833433/103595982-ab986000-4eb1-11eb-8c57-4726261b0a88.png">
> 
> ## Training Settings
> Before modifying anything, **first train with default settings to establish a performance baseline**. A full list of train.py settings can be found in the [train.py](https://github.com/ultralytics/yolov5/blob/master/train.py) argparser.
> 
> * **Epochs.** Start with 300 epochs. If this overfits early then you can reduce epochs. If overfitting does not occur after 300 epochs, train longer, i.e. 600, 1200 etc epochs.
> * **Image size.** COCO trains at native resolution of `--img 640`, though due to the high amount of small objects in the dataset it can benefit from training at higher resolutions such as `--img 1280`. If there are many small objects then custom datasets will benefit from training at native or higher resolution. Best inference results are obtained at the same `--img` as the training was run at, i.e. if you train at `--img 1280` you should also test and detect at `--img 1280`.
> * **Batch size.** Use the largest `--batch-size` that your hardware allows for. Small batch sizes produce poor batchnorm statistics and should be avoided.
> * **Hyperparameters.** Default hyperparameters are in [hyp.scratch.yaml](https://github.com/ultralytics/yolov5/blob/master/data/hyp.scratch.yaml). We recommend you train with default hyperparameters first before thinking of modifying any. In general, increasing augmentation hyperparameters will reduce and delay overfitting, allowing for longer trainings and higher final mAP. Reduction in loss component gain hyperparameters like `hyp['obj']` will help reduce overfitting in those specific loss components. For an automated method of optimizing these hyperparameters, see our [Hyperparameter Evolution Tutorial](https://docs.ultralytics.com/yolov5/tutorials/hyperparameter_evolution).
> 
> ## Further Reading
> If you'd like to know more a good place to start is Karpathy's 'Recipe for Training Neural Networks', which has great ideas for training that apply broadly across all ML domains:
> http://karpathy.github.io/2019/04/25/recipe/

Hello,

Thank you very much. 👍✔

**Comment by leeyunhome** [comment]
> @leeyunhome yes, your first training seems fine but too short. Double the epochs and retrain for better val loss and mAP. I'll post our full guide for improving results below.
> 
> 👋 Hello! Thanks for asking about improving training results. Most of the time good results can be obtained with no changes to the models or training settings, **provided your dataset is sufficiently large and well labelled**. If at first you don't get good results, there are steps you might be able to take to improve, but we always recommend users first train with all default settings before considering any changes. This helps establish a performance baseline and spot areas for improvement.
> 
> If you have questions about your training results **we recommend you provide the maximum amount of information possible** if you expect a helpful response, including results plots (train losses, val losses, P, R, mAP), PR curve, confusion matrix, training mosaics, test results and dataset statistics images such as labels.png. All of these are located in your `project/name` directory, typically `yolov5/runs/train/exp`.
> 
> We've put together a full guide for users looking to get the best results on their YOLOv5 trainings below.
> 
> ## Dataset
> * **Images per class.** ≥1.5k images per class
> * **Instances per class.** ≥10k instances (labeled objects) per class total
> * **Image variety.** Must be representative of deployed environment. For real-world use cases we recommend images from different times of day, different seasons, different weather, different lighting, different angles, different sources (scraped online, collected locally, different cameras) etc.
> * **Label consistency.** All instances of all classes in all images must be labelled. Partial labelling will not work.
> * **Label accuracy.** Labels must closely enclose each object. No space should exist between an object and it's bounding box. No objects should be missing a label.
> * **Background images.** Background images are images with no objects that are added to a dataset to reduce False Positives (FP). We recommend about 0-10% background images to help reduce FPs (COCO has 1000 background images for reference, 1% of the total).
> 
> [<img alt="COCO Analysis" width="800" src="https://user-images.githubusercontent.com/26833433/109398377-82b0ac00-78f1-11eb-9c76-cc7820669d0d.png">](https://arxiv.org/abs/1405.0312)
> 
> ## Model Selection
> Larger models like YOLOv5x will produce better results in nearly all cases, but have more parameters and are slower to run. For mobile applications we recommend YOLOv5s/m, for cloud or desktop applications we recommend YOLOv5l/x. See our README [table](https://github.com/ultralytics/yolov5#pretrained-checkpoints) for a full comparison of all models.
> 
> To start training from pretrained weights simply pass the name of the model to the `--weights` argument. Models download automatically from the [latest YOLOv5 release](https://github.com/ultralytics/yolov5/releases).
> 
> ```shell
> python train.py --data custom.yaml --weights yolov5s.pt
>                                              yolov5m.pt
>                                              yolov5l.pt
>                                              yolov5x.pt
> ```
> 
> <img alt="YOLOv5 Models" width="700" src="https://user-images.githubusercontent.com/26833433/103595982-ab986000-4eb1-11eb-8c57-4726261b0a88.png">
> 
> ## Training Settings
> Before modifying anything, **first train with default settings to establish a performance baseline**. A full list of train.py settings can be found in the [train.py](https://github.com/ultralytics/yolov5/blob/master/train.py) argparser.
> 
> * **Epochs.** Start with 300 epochs. If this overfits early then you can reduce epochs. If overfitting does not occur after 300 epochs, train longer, i.e. 600, 1200 etc epochs.
> * **Image size.** COCO trains at native resolution of `--img 640`, though due to the high amount of small objects in the dataset it can benefit from training at higher resolutions such as `--img 1280`. If there are many small objects then custom datasets will benefit from training at native or higher resolution. Best inference results are obtained at the same `--img` as the training was run at, i.e. if you train at `--img 1280` you should also test and detect at `--img 1280`.
> * **Batch size.** Use the largest `--batch-size` that your hardware allows for. Small batch sizes produce poor batchnorm statistics and should be avoided.
> * **Hyperparameters.** Default hyperparameters are in [hyp.scratch.yaml](https://github.com/ultralytics/yolov5/blob/master/data/hyp.scratch.yaml). We recommend you train with default hyperparameters first before thinking of modifying any. In general, increasing augmentation hyperparameters will reduce and delay overfitting, allowing for longer trainings and higher final mAP. Reduction in loss component gain hyperparameters like `hyp['obj']` will help reduce overfitting in those specific loss components. For an automated method of optimizing these hyperparameters, see our [Hyperparameter Evolution Tutorial](https://docs.ultralytics.com/yolov5/tutorials/hyperparameter_evolution).
> 
> ## Further Reading
> If you'd like to know more a good place to start is Karpathy's 'Recipe for Training Neural Networks', which has great ideas for training that apply broadly across all ML domains:
> http://karpathy.github.io/2019/04/25/recipe/

Hello,

1. Wouldn't it be difficult to expect accuracy at 320 resolution to train small objects like numbers on a license plate?

Is there a recommended ratio between object and image size?

2. If overfitting does not occur after 300 epochs, train longer, i.e. 600, 1200 etc epochs
=> He told me to increase the epoch if overfitting does not occur.
Do I need to train 600 and 1200 epochs again by designating the model trained with 300 epochs as the pretrained model?
if not
Do I need to train 600, 1200 epochs again by specifying yolov5.pt as a pre-trained model?

3. Why should I start with the epoch 300?

Thank you.

**Comment by glenn-jocher** [comment]
@leeyunhome for small objects use larger image sizes as the guidelines above state.

You'll have to experiment yourself to see what works best on your dataset.

**Comment by saktheeswaranswan** [comment]
> @leeyunhome for small objects use larger image sizes as the guidelines above state.
> 
> You'll have to experiment yourself to see what works best on your dataset.

I don't know why library like automl which can be targeted for specific metric like precision recall AUC can auto search and tell best accuracy can not be deployed eventough yolov5 and yolov7 are written in python if possible please tell
